src/IssueViz.Security/Taint/TaintStore.cs

```diff
@@ -39,21 +40,85 @@ internal interface ITaintStore : IIssuesStore
         /// Returns additional analysis information for the existing visualizations in the store.
         /// </summary>
         AnalysisInformation GetAnalysisInformation();
+
+        /// <summary>
+        /// Add the given issue to the existing list of visualizations.
+        /// If <see cref="GetAnalysisInformation"/> is null, the operation is ignored.
+        /// </summary>
+        void Add(IAnalysisIssueVisualization issueVisualization);
+
+        /// <summary>
+        /// Removes an issue with the given key from the existing list of visualizations.
+        /// If no matching issue is found, the operation is ignored.
+        /// </summary>
+        void Remove(string issueKey);
     }
 
     [Export(typeof(ITaintStore))]
     [Export(typeof(IIssuesStore))]
     [PartCreationPolicy(CreationPolicy.Shared)]
     internal sealed class TaintStore : ITaintStore
     {
-        private IAnalysisIssueVisualization[] taintVulnerabilities = Array.Empty<IAnalysisIssueVisualization>();
+        public event EventHandler<IssuesChangedEventArgs> IssuesChanged;
+
+        private static readonly object Locker = new object();
+
+        private List<IAnalysisIssueVisualization> taintVulnerabilities = Array.Empty<IAnalysisIssueVisualization>().ToList();
         private AnalysisInformation analysisInformation;
 
-        public IReadOnlyCollection<IAnalysisIssueVisualization> GetAll() => taintVulnerabilities;
-        
+        public IReadOnlyCollection<IAnalysisIssueVisualization> GetAll()
+        {
+            lock (Locker)
+            {
+                return taintVulnerabilities;
+            }
+        }
+
         public AnalysisInformation GetAnalysisInformation() => analysisInformation;
 
-        public event EventHandler<IssuesChangedEventArgs> IssuesChanged;
+        public void Add(IAnalysisIssueVisualization issueVisualization)
+        {
+            if (issueVisualization == null)
+            {
+                throw new ArgumentNullException(nameof(issueVisualization));
+            }
+
+            if (analysisInformation == null)
+            {
+                return;
+            }
+
+            lock (Locker)
+            {
+                taintVulnerabilities.Add(issueVisualization);
```

Comment:
Is it possible that we already have this issue viz in the store?
Perhaps worth asserting that we don't.